packages/eslint-plugin/src/rules/require-types-exports.ts

```diff
@@ -0,0 +1,408 @@
+import { TSESTree } from '@typescript-eslint/utils';
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+
+import { createRule } from '../util';
+import { DefinitionType } from '@typescript-eslint/scope-manager';
+
+type MessageIds = 'requireTypeExport';
+
+type FunctionNode =
+  | TSESTree.FunctionDeclaration
+  | TSESTree.TSDeclareFunction
+  | TSESTree.ArrowFunctionExpression
+  | TSESTree.FunctionExpression;
+
+export default createRule<[], MessageIds>({
+  name: 'require-types-exports',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      recommended: 'strict',
+      requiresTypeChecking: true,
+      description:
+        'Require exporting types that are used in exported functions declarations',
+    },
+    messages: {
+      requireTypeExport: 'Expected type "{{ name }}" to be exported',
+    },
+    schema: [],
+  },
+  defaultOptions: [],
+  create(context) {
+    const exportedTypes = new Set<string>();
+    const reported = new Set<string>();
+
+    function collectExportedTypes(program: TSESTree.Program): void {
```

Comment:
You also need to check for imported types:
```ts
import Arg from './types';

// This shouldn't report
export function f(arg: Arg): void {}
```

Reply:
Good catch, fixed, thanks! 😄